### PR TITLE
Support UTF-8 in metric creation, parsing, and exposition

### PR DESCRIPTION
**Pull request by ywwg**
Adding support for UTF-8 required substantial reworking of the parsing code because the previous code relied on regexes and .index calls that would incorrectly find characters inside quote marks. Note that this means the previous code was already broken for parsing exposition text where characters like braces and hashes were in label values. I have added tests to exercise the edge cases I can think of.

Does not address content negotiation, which will be a followon.

part of https://github.com/prometheus/client_python/issues/1013

### CODE REVIEW

File: tests/test_core.py [hunk 1]
```diff
@@ -663,6 +675,12 @@ def test_invalid_names_raise(self):
         self.assertRaises(ValueError, Counter, 'c_total', '', labelnames=['a:b'])
         self.assertRaises(ValueError, Counter, 'c_total', '', labelnames=['__reserved'])
         self.assertRaises(ValueError, Summary, 'c_total', '', labelnames=['quantile'])
+        disable_legacy_validation()
+        self.assertRaises(ValueError, Counter, '', 'help')
```
Comment (ywwg) [inline tests/test_core.py:679]: need to doublecheck these

**Comment by ywwg** [comment]
I'll squash commits when we're done

File: prometheus_client/validation.py [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+    _legacy_validation = False
+
+
+def enable_legacy_validation():
+    """Enable legacy name validation instead of allowing all UTF8 characters."""
+    global _legacy_validation
+    _legacy_validation = True
+
+
+def validate_metric_name(name: str) -> None:
```
Comment (csmarchbanks) [inline prometheus_client/validation.py:34]: Do we want these to be public now? Might be worth having them still prefixed with `_` but I also think they could be useful general functions for other libraries to use.

**Review by csmarchbanks — COMMENTED** [review]
I gave this a first runthrough. Generally looking ok to me. A couple thoughts:
1. If you want to split it up I think refactoring validation + removing regex parsing could be good opportunities. Plus have some tests that catch the incorrect behavior you mentioned (e.g. curly braces in a label value).
   - Tests for this would be nice in this PR too
2. Some more tests for various permutations would be nice, specifically there are not many tests for escaped label names/values right now

File: tests/openmetrics/test_parser.py [hunk 3]
```diff
@@ -187,17 +207,53 @@ def test_native_histogram(self):
...
+       
+        hfm = HistogramMetricFamily("native{histogram", "Is a basic example of a native histogram")
+        hfm.add_sample("native{histogram", {'xx{} # {}': ' EOF # {}}}'}, None, None, None, NativeHistogram(24, 100, 0, 0.001, 4, (BucketSpan(0, 2), BucketSpan(1, 2)), (BucketSpan(0, 2), BucketSpan(1, 2)), (2, 1, -3, 3), (2, 1, -2, 3)))
+        self.assertEqual([hfm], families)
 
     def test_native_histogram_with_labels(self):
         families = text_string_to_metric_families("""# TYPE hist_w_labels histogram
 # HELP hist_w_labels Is a basic example of a native histogram with labels
-hist_w_labels{foo="bar",baz="qux"} {count:24,sum:100,schema:0,zero_threshold:0.001,zero_count:4,positive_spans:[0:2,1:2],negative_spans:[0:2,1:2],positive_deltas:[2,1,-3,3],negative_deltas:[2,1,-2,3]}
+hist_w_labels{foo="bar",baz="qux"} {count:24,sum:100,schema:0,zero_threshold:0.001,zero_count:4,positive_spans:[0:2,1:2],negative_spans:[0:2,1:2],positive_deltas:[2,1,-3,3],negative_deltas:[2,1,-2,2]}
```
[the remark this answers is not in the document]
Comment (ywwg) [reply tests/openmetrics/test_parser.py:238]: oh hah, I changed this to make the test fail so I could see how it worked

File: prometheus_client/validation.py (hunk 2 above)
Comment (ywwg) [reply prometheus_client/validation.py:34]: the functions are used by both prometheus_client and prometheus_client/openmetrics so I thought they had to be public

Comment (csmarchbanks) [reply prometheus_client/validation.py:34]: Ahh, yeah Python doesn't really do public/private, it's just a convention.

Comment (ywwg) [reply prometheus_client/validation.py:34]: I am confused what change you want me to make, then -- leave it?  prepend _? something else?

**Comment by ywwg** [comment]
Added a test, fixed an issue using  .split

File: prometheus_client/parser.py [hunk 4]
```diff
@@ -45,54 +49,161 @@ def _is_character_escaped(s: str, charpos: int) -> bool:
...
+        if maxsplit > 0 and len(tokens) > maxsplit:
+            tokens[-1] = text[x:]
+            break
+        tokens[-1] = text[x:split_pos]
+        x = split_pos + 1
+        tokens.append('')
+    return tokens
+
+
+def unquote_unescape(text):
```
Comment (ywwg) [inline prometheus_client/parser.py:193]: maybe these should be private too?

File: prometheus_client/openmetrics/exposition.py [hunk 5]
```diff
@@ -59,16 +69,40 @@ def generate_latest(registry):
...
+    return '"{}"'.format(escape(s))
+
+
+def escape_label_name(s: str) -> str:
+    if _is_valid_legacy_labelname(s):
+        return s
+    return '"{}"'.format(escape(s))
+
+
+def escape(s: str) -> str:
```
Comment (csmarchbanks) [inline prometheus_client/openmetrics/exposition.py:107]: This one as private could be nice s well. I like the other escapes for metric and label names being public though as part of this work.

File: prometheus_client/parser.py (hunk 4 above)
Comment (csmarchbanks) [reply prometheus_client/parser.py:193]: I would say yes, thanks!

File: prometheus_client/parser.py [hunk 6]
```diff
@@ -104,34 +215,50 @@ def _parse_value_and_timestamp(s: str) -> Tuple[float, Optional[float]]:
...
-    value = float(values[0])
-    timestamp = (float(values[-1]) / 1000) if len(values) > 1 else None
+    value = parse_value(values[0])
+    timestamp = (parse_value(values[-1]) / 1000) if len(values) > 1 else None
     return value, timestamp
 
 
-def _parse_sample(text: str) -> Sample:
-    # Detect the labels in the text
+def parse_value(value):
```
Comment (csmarchbanks) [inline prometheus_client/parser.py:223]:
```suggestion
def _parse_value(value):
```
here as well to match `_parse_sample`

**Review by csmarchbanks — COMMENTED** [review]
Just a bit of prepending with more `_` and I think this is good to go! Plus cleaning up the commit history and such.

File: prometheus_client/openmetrics/exposition.py [hunk 7]
```diff
@@ -59,16 +71,47 @@ def generate_latest(registry):
...
         except Exception as exception:
             exception.args = (exception.args or ('',)) + (metric,)
             raise
 
     output.append('# EOF\n')
     return ''.join(output).encode('utf-8')
+
+
+def escape_metric_name(s: str) -> str:
+    """Escapes the metric name and puts it in quotes iff the name does not
```
Comment (vesari) [inline prometheus_client/openmetrics/exposition.py:98]:
```suggestion
    """Escapes the metric name and puts it in quotes if the name does not
```

File: prometheus_client/openmetrics/exposition.py [hunk 8]
```diff
@@ -59,16 +71,47 @@ def generate_latest(registry):
...
+    conform to the legacy Prometheus character set.
+    """
+    if _is_valid_legacy_metric_name(s):
+        return s
+    return '"{}"'.format(_escape(s))
+
+
+def escape_label_name(s: str) -> str:
+    """Escapes the label name and puts it in quotes iff the name does not
```
Comment (vesari) [inline prometheus_client/openmetrics/exposition.py:107]:
```suggestion
    """Escapes the label name and puts it in quotes if the name does not
```

**Review by vesari — APPROVED** [review]
Thanks for this great work, LGTM!

File: prometheus_client/openmetrics/exposition.py (hunk 7 above)
Comment (ywwg) [reply prometheus_client/openmetrics/exposition.py:98]: This is shorthand for "if and only if" which is perhaps obscure, but adds detail to the explanation.  I can write out the full phrase if that's better

Comment (vesari) [reply prometheus_client/openmetrics/exposition.py:98]: Oops, my bad, please leave them as they are :)

**Review by csmarchbanks — APPROVED** [review]
